sympy/physics/quantum/lapply.py

```diff
@@ -0,0 +1,1419 @@
+   to * resp. Mul. Special rules have been provided for handling of powers
+   with idempotent or involutoric factors in the base.
+
+Object type specifics have completely been deployed to @dispatch handlers:
+a. The algorithm that walks the expression tree and determines to pair factors
+   a,b,c,d as (a*b)*(c*d) or a*((b*c)*d) etc. is completely object type agnostic
+   (lapply_Mul2). Walking of the tree is dependant on the success of pairing:
+   I decided to implement (a*(b*(c*d))) association from the right in general
+   but depending on whether factors interacted to new factors and re-tries to
+   the right side in that case.
+b. Recursing into inner structure of an object (lapply1_type handlers)
+c. Splitting up composite objects in commutative (c) and non-commutative
+   (nc) factors (c_nc_ncef handlers)
+d. All rules how to multiply two object types (lapply_mul2elems with
+   lapply_mul2types_x handlers)
+
+Moreover:
+5. expand() is not used indiscriminately: Only do an expansion of Add if
+   required to apply operators, so keeping expansion to the bare minimum. As
+   default, distribute commutative factors over sums as this allows terms to
+   cancel out thus simplifying the expression. In order to simplify commuta
+   -tive factors lapply() supports the three major hints of expand() "mul",
+   "power_base" and "power_exp" (see options mul, power_base and power_exp).
+6. apply recursion along the structure of the expression but don't use
+   recursion for iterations: this avoids hitting the Python recursion limit
+   in very complex expressions or powers with exponents > 1200 or so.
+7. No need to try to maintain subclass types of objects and expressions, as
+   we simply cannot know whether a product is still in a subclass of Mul etc.
+   So use the top types.
+"""
+
+from typing import cast, Any, List, Tuple, Callable, Union  # For Python 3.8
+from enum import Enum
+
+from sympy.core.expr import Expr
+from sympy.core.add import Add
+from sympy.core.mul import Mul
+from sympy.core.power import Pow
+from sympy.core.singleton import S
+from sympy.core.numbers import Number
+from sympy.functions.elementary.integers import floor
+from sympy.multipledispatch.dispatcher import Dispatcher
+
+# lapply is object agnostic. All object specific information has been
+# isolated in handlers that are dispatched by object type.
+# Imports required to define object types for handlers are coded
+# just in front of the handlers.
+
+
+__all__ = [
+    'lapply'
+]
+
+#--------+---------+---------+---------+---------+---------+---------+---------
+# lapply function header and doc string
+#
+# Options taken from expand() have same names (mul, power_xxx, tensorproduct)
+#------------------------------------------------------------------------------
+
+def lapply(e:Expr, mul=True, Add=Add,
+                   power_base=None, power_exp=False, nested_exp=None,
+                   apply_exp=False, cache_for_pow=None, **options) -> Expr:

```

Comment:
The `**options `are required for options that users want to pass on to their self-written handlers and that lapply doesn't know about, so their keywords cannot be listed in the function head. The `original qapply` had this feature and passed `options `on to calls of `._apply_operator`. Removing this would be a breaking change. 

I would advocate to leave this feature in as it gives users flexibility for their self-written objects and handlers, and accept the downside that misspelled options are not flagged.

An idea here: What if all user-defined options had to be capitalized? That might still be a  breaking change, but probably marginal, but most misspelled options could be flagged then.